util-images/network/netperfbenchmark/pkg/worker/util.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package worker
+
+import (
+	"errors"
+	"math"
+	"net/http"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"k8s.io/klog"
+)
+
+// Protocols supported
+const (
+	ProtocolTCP  = "TCP"
+	ProtocolUDP  = "UDP"
+	ProtocolHTTP = "HTTP"
+)
+
+// Number of metrics for each protocol
+const (
+	tcpMetricsCount  = 2
+	udpMetricsCount  = 11
+	httpMetricsCount = 12
+)
+
+// Iperf results vary from protocol,required for proper parsing
+const (
+	zeroFormatTCP         = "0.0"
+	fractionZeroFormatTCP = ".0"
+	zeroFormatUDP         = "0.00"
+	fractionZeroFormatUDP = ".00"
+)
+
+// Function to be applied for each metric for aggregation
+var (
+	iperfUDPFunction = []string{"", "", "", "Sum", "Sum", "Sum", "Sum", "Sum", "Avg", "Avg", "Min", "Max", "Avg", "Sum"}
+	iperfTCPFunction = []string{"", "", "", "Sum", "Avg"}
+)
+
+// parseResult parses the response received for each protocol type
+func parseResult(protocol string, result []string, testDuration string) ([]float64, error) {
+	switch protocol {
+	case ProtocolTCP:
+		return parseIperfResponse(result, testDuration, ProtocolTCP, tcpMetricsCount, zeroFormatTCP, fractionZeroFormatTCP, iperfTCPFunction)
+	case ProtocolUDP:
+		return parseIperfResponse(result, testDuration, ProtocolUDP, udpMetricsCount, zeroFormatUDP, fractionZeroFormatUDP, iperfUDPFunction)
+	case ProtocolHTTP:
+		return parseSiegeResponse(result)
+	default:
+		return nil, errors.New("invalid protocol: " + protocol)
+	}
+}
+
+func parseIperfResponse(result []string, testDuration string, protocol string, metricCount int, zeroFormat string, fractionZeroFormat string, operators []string) ([]float64, error) {
+	aggregatedResult := make([]float64, 0, metricCount)
+	unitRegex := regexp.MustCompile(`%|\[\s+|\]\s+|KBytes\s+|KBytes/sec\s*|sec\s+|pps\s*|ms\s+|/|\(|\)\s+`)
+	multiSpaceRegex := regexp.MustCompile(`\s+`)
+	hyphenSpaceRegex := regexp.MustCompile(`\-\s+`)
+	count := 0
+	sessionID := make(map[string]bool)
+	for _, row := range result {
+		formattedString := hyphenSpaceRegex.ReplaceAllString(multiSpaceRegex.ReplaceAllString(unitRegex.ReplaceAllString(row, " "), " "), "-")
+		if !strings.Contains(formattedString, zeroFormat+"-"+testDuration+fractionZeroFormat) {
+			continue
+		}
+		split := strings.Split(formattedString, " ")
```

Comment:
Can you encapsulate the above 5 lines or so into a separate functIon?

Like parseIperfLine or sth like that?

Reply:
Modified.